runtime/src/main/java/org/corfudb/runtime/clients/BaseHandler.java

```diff
@@ -159,4 +137,176 @@ private static Object handleWrongClusterId(CorfuPayloadMsg<WrongClusterMsg> msg,
+    }
+
+    /**
+     * Handle a UNKNOWN_ERROR response from the server.
+     *
+     * @param msg The wrong epoch message
+     * @param ctx The context the message was sent under
+     * @param r   A reference to the router
+     * @return none, throw a wrong epoch exception instead.
+     */
+    @ServerErrorsHandler(type = ErrorCase.UNKNOWN_ERROR)
+    private static Object handleUnknownError(ResponseMsg msg, ChannelHandlerContext ctx,
+                                             IClientRouter r) throws Throwable {
+        ByteString bs = msg.getPayload().getServerError().getUnknownError().getThrowable();
+        byte[] bytes = new byte[bs.size()];
+        bs.copyTo(bytes, 0);
+        try (ByteArrayInputStream bis = new ByteArrayInputStream(bytes)) {
+            try (ObjectInputStream ois = new ObjectInputStream(bis)) {
+                throw (Throwable) ois.readObject();
+            }
+        }
+    }
+
+    /**
+     * Handle a WRONG_EPOCH_ERROR response from the server.
+     *
+     * @param msg The wrong epoch message
+     * @param ctx The context the message was sent under
+     * @param r   A reference to the router
+     * @return none, throw a wrong epoch exception instead.
+     */
+    @ServerErrorsHandler(type = ErrorCase.WRONG_EPOCH_ERROR)
+    private static Object handleWrongEpochError(ResponseMsg msg, ChannelHandlerContext ctx,
+                                                IClientRouter r) {
+        long correctEpoch = msg.getPayload().getServerError().getWrongEpochError().getCorrectEpoch();
+        throw new WrongEpochException(correctEpoch);
+    }
+
+    /**
+     * Handle a NOT_READY_ERROR response from the server.
+     *
+     * @param msg The wrong epoch message
+     * @param ctx The context the message was sent under
+     * @param r   A reference to the router
+     * @return none, throw a wrong epoch exception instead.
+     */
+    @ServerErrorsHandler(type = ErrorCase.NOT_READY_ERROR)
+    private static Object handleNotReadyError(ResponseMsg msg, ChannelHandlerContext ctx,
+                                              IClientRouter r) {
+        throw new ServerNotReadyException();
+    }
+
+    /**
+     * Handle a WRONG_CLUSTER_ERROR response from the server.
+     *
+     * @param msg The wrong epoch message
+     * @param ctx The context the message was sent under
+     * @param r   A reference to the router
+     * @return none, throw a wrong epoch exception instead.
+     */
+    @ServerErrorsHandler(type = ErrorCase.WRONG_CLUSTER_ERROR)
+    private static Object handleWrongClusterError(ResponseMsg msg, ChannelHandlerContext ctx,
+                                                  IClientRouter r) {
+        WrongClusterErrorMsg errorMsg = msg.getPayload().getServerError().getWrongClusterError();
+        UUID expectedCluster = CorfuProtocolCommon.getUUID(errorMsg.getExpectedClusterId());
+        UUID actualCluster = CorfuProtocolCommon.getUUID(errorMsg.getProvidedClusterId());
```

Comment:
I believe this is backwards. When the server generates this error, the `expectedClusterId` is its own, and the `providedClusterId` is the id provided by the client obtained from the header. However, when the client throws this exception, the `expectedCluster` is its own and the `actualCluster` is that of the server.